lib/route/cls/flower.c

```diff
@@ -598,6 +670,124 @@ int rtnl_flower_get_ip_dscp(struct rtnl_cls *cls, uint8_t *dscp, uint8_t *mask)
 	return 0;
 }
 
+/**
+ * Set IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_src(struct rtnl_cls *cls, int addr, int mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_src = nl_addr_build(AF_INET,
+					       (unsigned char *)&addr,
+					       sizeof(__be32));
+
+		f->cf_mask |= FLOWER_ATTR_IPV4_SRC;
+
+		if (mask) {
+			uint32_t m = 0xFFFFFFFF << (32 - mask);
+			f->cf_ipv4_src_mask = htonl(m);
+			f->cf_mask |= FLOWER_ATTR_IPV4_SRC_MASK;
+		}
+
+		return 0;
+	}
+
+	return -NLE_FAILURE;
+}
+
+/**
+ * Get IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_get_ipv4_src(struct rtnl_cls *cls, struct nl_addr *addr,
+			     uint32_t *mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data_peek(TC_CAST(cls))))
+		return -NLE_INVAL;
+
+	if (!(f->cf_mask & FLOWER_ATTR_IPV4_SRC))
+		return -NLE_MISSING_ATTR;
+
+	if (addr)
+		memcpy(addr, f->cf_ipv4_src, AF_INET);
+
+	*mask = f->cf_ipv4_src_mask;
+
+	return 0;
+}
+
+/**
+ * Set IPv4 destination address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 destination address
+ * @arg mask		mask for IPv4 destination address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_dst(struct rtnl_cls *cls, int addr, int mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_dst = nl_addr_build(AF_INET,
+					       (unsigned char *)&addr,
+					       sizeof(__be32));
+
+		f->cf_mask |= FLOWER_ATTR_IPV4_DST;
+
+		if (mask) {
+			uint32_t m = 0xFFFFFFFF << (32 - mask);
+			f->cf_ipv4_dst_mask = htonl(m);
+			f->cf_mask |= FLOWER_ATTR_IPV4_DST_MASK;
+		}
+
+		return 0;
+	}
+
+	return -NLE_FAILURE;
+}
+
+/**
+ * Get IPv4 destination address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 destination address
+ * @arg mask		mask for IPv4 destination address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_get_ipv4_dst(struct rtnl_cls *cls, struct nl_addr *addr,
+			     uint32_t *mask)
```

Comment:
it would seem that the IPv4 is always a `uint32_t` (actually `in_addr_t` or `struct in_addr`, see above).

This is not something that could change, and the function name (`rtnl_flower_get_ipv4_dst()`) also ties that down. I think it's needlessly complicated to expose the address as `struct nl_addr`, in particular, since `mask` is already a `uint32_t`. I think the function should be
```C
int rtnl_flower_get_ipv4_dst(struct rtnl_cls *cls, in_addr_t *out_addr, in_addr_t *out_mask);
```